### PR TITLE
Feedback

### PR DESCRIPTION
**Pull request by Muscraft**

### CODE REVIEW

File: examples/footer.svg
Comment (epage) [inline examples/footer.svg:1]: `note:` doesn't get colored?

File: examples/format.svg
Comment (epage) [inline examples/format.svg:1]: Changes seem to be
- We now bold the `:` in `error:`
- We will no longer color the ` ` (whitespace) between the column and `|`

File: examples/multislice.svg
Comment (epage) [inline examples/multislice.svg:1]: We added a blank line between the first snippet and `:::`

File: examples/expected_type.svg
Comment (epage) [inline examples/expected_type.svg:1]: We are now folding lines between annotations

File: src/renderer/mod.rs [hunk 1]
```diff
@@ -10,20 +10,29 @@
...
+use crate::renderer::source_map::{AnnotatedLineInfo, Loc, SourceMap};
+use crate::renderer::styled_buffer::StyledBuffer;
 use crate::snippet::Message;
+use crate::{Level, Snippet};
 pub use anstyle::*;
-use display_list::DisplayList;
+use indexmap::IndexMap;
 use margin::Margin;
-use std::fmt::Display;
+use rustc_hash::{FxHashMap, FxHasher};
```
Comment (epage) [inline src/renderer/mod.rs]: should we consider making this an optional feature or removing it?

File: src/snippet.rs [hunk 2]
```diff
@@ -10,12 +10,20 @@
...
+pub(crate) const HELP_TXT: &str = "help";
+pub(crate) const INFO_TXT: &str = "info";
+pub(crate) const NOTE_TXT: &str = "note";
+pub(crate) const WARNING_TXT: &str = "warning";
+
 /// Primary structure provided for formatting
 ///
 /// See [`Level::title`] to create a [`Message`]
-#[derive(Debug)]
+#[derive(Clone, Debug)]
```
Comment (epage) [inline src/snippet.rs]: Could you split API changes like this into a commit before hand?

File: examples/format.svg
Comment (epage) [reply examples/format.svg:1]: If its not too much trouble, could this be hacked in to the old renderer in a commit before so the test changes are easier to see?

File: examples/multislice.svg
Comment (epage) [inline examples/multislice.svg:1]: > refactor: Right align line numbers 

That is not a refactor if its user visible

File: benches/bench.rs
Comment (epage) [inline benches/bench.rs:1]: > refactor: Move to new API 

its not a refactor if its user visible.

This is also a breaking change and should be called out

File: benches/bench.rs [hunk 3]
```diff
@@ -24,20 +24,22 @@ fn simple() -> String {
...
-                    .span(5..19)
-                    .label("expected `Option<String>` because of return type"),
-            )
-            .annotation(
-                Level::Error
-                    .span(26..724)
-                    .label("expected enum `std::option::Option`"),
-            ),
+    let message = Level::Error.message("mismatched types").id("E0308").group(
+        Group::new().element(
```
Comment (epage) [inline benches/bench.rs:28]: Could you explain the new API in the commit or PR?

For example, why was a `Group` container chosen over `group` accepting a `Vec` or `IntoIterator`

File: src/snippet.rs [hunk 4]
```diff
@@ -20,16 +10,23 @@ pub(crate) const INFO_TXT: &str = "info";
...
+        memchr::memchr_iter(b'\n', body.as_bytes())
+            .count()
+            .saturating_sub(1)
+    }
+    #[cfg(not(feature = "simd"))]
+    {
+        body.lines().count().saturating_sub(1)
+    }
+}
+
```
Comment (epage) [inline src/snippet.rs]: Generally, implementation details should be at the end.

File: src/snippet.rs [hunk 5]
```diff
@@ -38,104 +35,178 @@ impl<'a> Message<'a> {
         self
     }
 
-    pub fn snippet(mut self, slice: Snippet<'a>) -> Self {
-        self.snippets.push(slice);
+    pub fn group(mut self, group: Group<'a>) -> Self {
```
Comment (epage) [inline src/snippet.rs:25]: One of the places where we deal with groups, we'll need to explain the difference between a single group vs many groups.  Unsure whether a suggested policy belongs in this package or somewhere else but at least Cargo and Rustc will need to be consistent

File: src/snippet.rs [hunk 6]
```diff
@@ -38,104 +35,178 @@ impl<'a> Message<'a> {
...
-            .snippets
-            .iter()
-            .map(|s| {
-                let start = s
-                    .annotations
-                    .iter()
-                    .map(|a| a.range.start)
-                    .min()
-                    .unwrap_or(0);
+    pub fn needs_col_separator(mut self, needs_col_separator: bool) -> Self {
```
Comment (epage) [inline src/snippet.rs]: What role does this serve?

File: src/snippet.rs [hunk 7]
```diff
@@ -149,32 +220,40 @@ impl<'a> Snippet<'a> {
...
-    pub fn annotation(mut self, annotation: Annotation<'a>) -> Self {
-        self.annotations.push(annotation);
+    pub fn fold(mut self, fold: bool) -> Self {
+        self.fold = fold;
         self
     }
 
-    pub fn annotations(mut self, annotation: impl IntoIterator<Item = Annotation<'a>>) -> Self {
-        self.annotations.extend(annotation);
+    pub fn col_separator(mut self, col_separator: bool) -> Self {
```
Comment (epage) [inline src/snippet.rs]: What is this?

File: src/snippet.rs [hunk 8]
```diff
@@ -184,57 +263,77 @@ impl<'a> Annotation<'a> {
...
-    /// Warning annotations are displayed using blue color and "-" character.
-    Warning,
-    Info,
-    Note,
-    Help,
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum AnnotationKind {
+    Primary, // in rustc terms, "primary"; color to message's level
+    Context, // in rustc terms, "secondary"; fixed color
+             // Level(Level),  // from old API, dropped
```
Comment (epage) [inline src/snippet.rs]: Please use doc comments

File: src/snippet.rs [hunk 9]
```diff
@@ -184,57 +263,77 @@ impl<'a> Annotation<'a> {
...
+
+    pub fn primary(mut self, primary: bool) -> Self {
+        self.primary = primary;
+        self
+    }
+
+    pub fn standalone(mut self, standalone: bool) -> Self {
+        self.standalone = standalone;
+        self
+    }
```
Comment (epage) [inline src/snippet.rs]: What are these?

File: src/snippet.rs [hunk 10]
```diff
@@ -184,57 +263,77 @@ impl<'a> Annotation<'a> {
...
+
+    pub fn line(mut self, line: usize) -> Self {
+        self.line = Some(line);
+        self
+    }
+
+    pub fn char_column(mut self, char_column: usize) -> Self {
+        self.char_column = Some(char_column);
+        self
+    }
```
Comment (epage) [inline src/snippet.rs:362]: These are new.  How do they relate to the rest of the API?

Should we make new parts of the API an extra commit on top of the existing API for easier comparison and  discussion?

File: src/snippet.rs [hunk 11]
```diff
@@ -149,32 +220,40 @@ impl<'a> Snippet<'a> {
...
 }
 
-/// An annotation for a [`Snippet`].
-///
-/// See [`Level::span`] to create a [`Annotation`]
+impl<'a> From<Snippet<'a, Annotation<'a>>> for Element<'a> {
+    fn from(value: Snippet<'a, Annotation<'a>>) -> Self {
+        Element::Cause(value)
+    }
+}
```
Comment (epage) [inline src/snippet.rs]: I'd put all of these `From`s next to `Element` as this is about defining how to interact with an `Element` and not each of the other types

File: src/snippet.rs [hunk 12]
```diff
@@ -284,6 +309,91 @@ impl AnnotationKind {
...
+    pub fn new(range: Range<usize>, replacement: &'a str) -> Self {
+        Self { range, replacement }
+    }
+
+    pub(crate) fn is_addition(&self, sm: &SourceMap<'_>) -> bool {
+        !self.replacement.is_empty() && !self.replaces_meaningful_content(sm)
+    }
+
+    pub(crate) fn is_deletion(&self, sm: &SourceMap<'_>) -> bool {
+        self.replacement.trim().is_empty() && self.replaces_meaningful_content(sm)
```
Comment (epage) [inline src/snippet.rs:288]: Why `trim` and only on this one?

File: tests/formatter.rs
Comment (epage) [inline tests/formatter.rs:1]: Could we add these tests in a prior commit, either without the patches or using annotations instead of patches?

File: src/renderer/mod.rs [hunk 13]
```diff
@@ -143,6 +145,11 @@ impl Renderer {
         self
     }
 
+    pub const fn theme(mut self, output_theme: OutputTheme) -> Self {
+        self.theme = output_theme;
+        self
+    }
+
```
Comment (epage) [inline src/renderer/mod.rs:147]: I feel like if I was looking for how to enable unicode support, I wouldn't be looking at theming.

File: tests/formatter.rs [hunk 14]
```diff
@@ -1690,3 +1691,424 @@ help: consider removing the `?Sized` bound to make the type parameter `Sized`
     let renderer = Renderer::plain();
     assert_data_eq!(renderer.render(input_new), expected);
 }
+
+#[test]
+fn e0271() {
```
Comment (epage) [inline tests/formatter.rs:1698]: what are these test names?

Can we add these tests without unicode support and then show how unicode support changes them?

File: tests/formatter.rs
Comment (epage) [inline tests/formatter.rs:1]: > chore: Align more with rustc's output 

"chore"s don't generally have user-visible impact. Isn't this a fix?

File: Cargo.toml
Comment (epage) [inline Cargo.toml:1]: >  refactor: Better match rustc's internal renderer 

Since this has end-user impact, this isn't just about better matching the internals but the result and that should be reflected in the commit type

Ideally, the commit description would cover expected changes, including regressions

File: tests/formatter.rs
Comment (epage) [reply tests/formatter.rs:1]: And in what way?

File: src/renderer/mod.rs
Comment (epage) [inline src/renderer/mod.rs:1]: > fix: Use display and char pos in draw_line

That says what its doing; not what its fixing
